pkg/controllers/nodeclaim/tagging/suite_test.go

```diff
@@ -0,0 +1,195 @@
+var env *coretest.Environment
+var taggingController controller.Controller
+
+func TestAPIs(t *testing.T) {
+	ctx = TestContextWithLogger(t)
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Tagging")
+}
+
+var _ = BeforeSuite(func() {
+	env = coretest.NewEnvironment(scheme.Scheme, coretest.WithCRDs(apis.CRDs...))
+	ctx = coresettings.ToContext(ctx, coretest.Settings())
+	ctx = settings.ToContext(ctx, test.Settings())
+	awsEnv = test.NewEnvironment(ctx, env)
+	taggingController = tagging.NewController(env.Client, awsEnv.InstanceProvider)
+})
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = BeforeEach(func() {
+	awsEnv.Reset()
+})
+
+var _ = AfterEach(func() {
+	ExpectCleanedUp(ctx, env.Client)
+})
+
+var _ = Describe("TaggingController", func() {
+	var ec2Instance *ec2.Instance
+
+	BeforeEach(func() {
+		ec2Instance = &ec2.Instance{
+			State: &ec2.InstanceState{
+				Name: aws.String(ec2.InstanceStateNameRunning),
+			},
+			Tags: []*ec2.Tag{
+				{
+					Key:   aws.String(fmt.Sprintf("kubernetes.io/cluster/%s", settings.FromContext(ctx).ClusterName)),
+					Value: aws.String("owned"),
+				},
+				{
+					Key:   aws.String(corev1beta1.NodePoolLabelKey),
+					Value: aws.String("default"),
+				},
+				{
+					Key:   aws.String(corev1beta1.ManagedByAnnotationKey),
+					Value: aws.String(settings.FromContext(ctx).ClusterName),
+				},
+			},
+			PrivateDnsName: aws.String(fake.PrivateDNSName()),
+			Placement: &ec2.Placement{
+				AvailabilityZone: aws.String(fake.DefaultRegion),
+			},
+			InstanceId:   aws.String(fake.InstanceID()),
+			InstanceType: aws.String("m5.large"),
+		}
+
+		awsEnv.EC2API.Instances.Store(*ec2Instance.InstanceId, ec2Instance)
+	})
+
+	It("shouldn't tag instances without a Node", func() {
+		nodeClaim := coretest.NodeClaim(corev1beta1.NodeClaim{
+			Status: corev1beta1.NodeClaimStatus{
+				ProviderID: fake.ProviderID(*ec2Instance.InstanceId),
+			},
+		})
+
+		ExpectApplied(ctx, env.Client, nodeClaim)
+		ExpectReconcileSucceeded(ctx, taggingController, client.ObjectKeyFromObject(nodeClaim))
+		Expect(nodeClaim.Annotations).To(Not(HaveKey(v1beta1.AnnotationInstanceTagged)))
+		Expect(lo.ContainsBy(ec2Instance.Tags, func(tag *ec2.Tag) bool {
+			return *tag.Key == tagging.TagName
+		})).To(BeFalse())
+	})
+
+	It("should't tag instance with malformed providerID", func() {
+		nodeClaim := coretest.NodeClaim(corev1beta1.NodeClaim{
+			Status: corev1beta1.NodeClaimStatus{
+				ProviderID: "Bad providerID",
+				NodeName:   "default",
+			},
+		})
+
+		ExpectApplied(ctx, env.Client, nodeClaim)
+		ExpectReconcileSucceeded(ctx, taggingController, client.ObjectKeyFromObject(nodeClaim))
+		Expect(nodeClaim.Annotations).To(Not(HaveKey(v1beta1.AnnotationInstanceTagged)))
+		Expect(lo.ContainsBy(ec2Instance.Tags, func(tag *ec2.Tag) bool {
+			return *tag.Key == tagging.TagName
+		})).To(BeFalse())
+	})
+
+	It("should gracefully handle missing NodeClaim", func() {
```

Comment:
I was thinking that the case where the instance doesn't exist might be more complex here. Should we check for that as well?

Reply:
This is actually a good callout from the test case. Should we ignore tagging instances from NodeClaims that are actively deleting? So check the DeletionTimestamp is set and ignore it if it is

Reply:
Both seem reasonable to me 👍